cmd/influxd/handler.go

```diff
@@ -1,42 +1,117 @@
 }
 
 // NewHandler returns a new instance of Handler.
-func NewHandler(bh, sh http.Handler) *Handler {
-	return &Handler{
-		brokerHandler: bh,
-		serverHandler: sh,
-	}
+func NewHandler() *Handler {
+	return &Handler{}
 }
 
 // ServeHTTP responds to HTTP request to the handler.
 func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	// Route raft and messaging paths to the broker.
-	if strings.HasPrefix(r.URL.Path, "/raft") || strings.HasPrefix(r.URL.Path, "/messaging") {
-		if h.brokerHandler == nil {
-			http.NotFound(w, r)
-			return
+	if strings.HasPrefix(r.URL.Path, "/raft") {
+		h.serveRaft(w, r)
+		return
+	}
+	if strings.HasPrefix(r.URL.Path, "/messaging") {
+		h.serveMessaging(w, r)
+		return
+	}
+
+	h.serveData(w, r)
+}
+
+// serveMessaging responds to broker requests
+func (h *Handler) serveMessaging(w http.ResponseWriter, r *http.Request) {
+	// If we're running a broker, handle the broker endpoints
+	if h.Broker != nil {
+		mh := &messaging.Handler{
+			Broker:      h.Broker.Broker,
+			RaftHandler: &raft.Handler{Log: h.Log},
 		}
+		mh.ServeHTTP(w, r)
+		return
+	}
+
+	if h.Server == nil {
+		log.Println("no broker or server configured to handle messaging endpoints")
+		http.Error(w, "server unavailable", http.StatusServiceUnavailable)
+		return
+	}
+
+	// Redirect to a valid broker to handle the request
+	h.redirect(h.Server.BrokerURLs(), w, r)
+}
 
-		h.brokerHandler.ServeHTTP(w, r)
+// serveRaft responds to raft requests.
+func (h *Handler) serveRaft(w http.ResponseWriter, r *http.Request) {
+	if h.Log != nil {
+		rh := raft.Handler{Log: h.Log}
+		rh.ServeHTTP(w, r)
 		return
 	}
 
-	// Route all other paths to the server.
-	if h.serverHandler == nil {
-		http.NotFound(w, r)
+	if h.Server == nil {
+		log.Println("no broker or server configured to handle messaging endpoints")
+		http.Error(w, "server unavailable", http.StatusServiceUnavailable)
 		return
 	}
-	h.serverHandler.ServeHTTP(w, r)
+
+	// Redirect to a valid broker to handle the request
+	h.redirect(h.Server.BrokerURLs(), w, r)
+}
+
+// serveData responds to data requests
+func (h *Handler) serveData(w http.ResponseWriter, r *http.Request) {
+	if h.Server != nil {
+		sh := httpd.NewHandler(h.Server, h.Config.Authentication.Enabled, version)
+		sh.WriteTrace = h.Config.Logging.WriteTracing
+		sh.ServeHTTP(w, r)
+		return
+	}
+
+	if h.Broker == nil {
+		log.Println("no broker or server configured to handle messaging endpoints")
+		http.Error(w, "server unavailable", http.StatusServiceUnavailable)
+		return
+
+	}
+
+	// Redirect to a valid data URL to handle the request
+	h.redirect(h.Broker.Topic(influxdb.BroadcastTopicID).DataURLs(), w, r)
+}
+
+// redirect redirects a request to URL in u.  If u is an empty slice,
+// a 503 is returned
+func (h *Handler) redirect(u []url.URL, w http.ResponseWriter, r *http.Request) {
+	// No valid URLs, return an error
+	if len(u) == 0 {
+		http.Error(w, "service unavailable", http.StatusServiceUnavailable)
```

Comment:
It is possible that a data node could ask a broker a data node question, and that specific broker did not know about any data nodes.  In that case, the urls would be empty, and you could return this error.  This should only happen with a mis-configured cluster.  However, since it is outside of our control, we have to check for it.